File: DynamicProgramming/FindTargetSumWays/findTargetSumWays.py
# 根据该网站学习并且编写: https://labuladong.gitbook.io/algo/dong-tai-gui-hua-xi-lie/targetsum

class FindTargetSumWays(object):
    def __init__(self):
        self.result = 0
        self.memo = dict()
        print('init...')

    # 回溯法
    def backtrack(self, nums, i, rest):
        if i == len(nums):
            if rest == 0:
                self.result += 1
            return

        self.backtrack(nums, i + 1, rest + nums[i])
        self.backtrack(nums, i + 1, rest - nums[i])

    def dp(self, nums, i, rest):
        if i == len(nums):
            if rest == 0:
                return 1
            return 0
        
        if (i, rest) in self.memo:
            return self.memo[(i, rest)]

        res = self.dp(nums, i + 1, rest - nums[i]) + self.dp(nums, i + 1, rest + nums[i])
        self.memo[(i, rest)] = res
        return res

ftsw = FindTargetSumWays()
ftsw.backtrack([1, 1, 1, 1, 1], 0, 3)
print(ftsw.result)
print(ftsw.dp([1, 1, 1, 1, 1], 0, 3))

File: DynamicProgramming/FindTargetSumWays/findTargetSumWays.js
// 根据该网站学习并且编写: https://labuladong.gitbook.io/algo/dong-tai-gui-hua-xi-lie/targetsum

class FindTargetSumWays {
    constructor() {
        this.result = 0
        this.memo = new Map()
    }

    // 回溯法
    backtrack(nums, i, rest) {
        if (i == nums.length) {
            if (rest == 0) {
                this.result++
            }
            return;
        }

        this.backtrack(nums, i + 1, rest + nums[i])
        this.backtrack(nums, i + 1, rest - nums[i])
    }

    // 备忘录法
    dp(nums, i, rest) {
        if (i == nums.length) {
            if (rest == 0) {
                return 1
            }
            return 0
        }

        let key = i + "," + rest
        if (this.memo.has(key)) {
            return this.memo.get(key)
        }

        let res = this.dp(nums, i + 1, rest + nums[i]) + this.dp(nums, i + 1, rest - nums[i])
        this.memo.set(key, res)
        return res
    }
}

let ftsw = new FindTargetSumWays()
let nums = [1, 1, 1, 1, 1]
ftsw.backtrack(nums, 0, 3)
console.log(ftsw.result)
console.log(ftsw.dp(nums, 0, 3))